llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp

```diff
@@ -1780,11 +1790,21 @@ void HWAddressSanitizer::instrumentGlobals() {
     if (GV.hasCommonLinkage())
       continue;
 
-    // Globals with custom sections may be used in __start_/__stop_ enumeration,
-    // which would be broken both by adding tags and potentially by the extra
-    // padding/alignment that we insert.
-    if (GV.hasSection())
-      continue;
+    if (ClAllGlobals) {
+      // Avoid adding metadata emitted for the hwasan instrumentation itself.
+      // Code which makes assumptions about memory layout of globals between
+      // __start_<section>/__end_<section> linker-generated symbols may need
+      // manual adaptation.
+      auto Section = GV.getSection();
+      if (Section == "hwasan_globals" || Section == ".note.hwasan.globals")
```

Comment:
Is this check even needed if you move the call to instrumentGlobals above createHwasanCtorComdat?

Reply:
I think it won't hurt to also skip the `llvm.metadata` section